bazel_codegen/lib/bazel_codegen.dart

```diff
@@ -0,0 +1,82 @@
+import 'dart:async';
+
+import 'package:build/build.dart';
+
+import 'src/arg_parser.dart';
+import 'src/run_phases.dart';
+
+typedef Builder BuilderFactory(List<String> args);
+
+/// Wrap [builder] as a [BuiderFactory] that expects no arguments.
+///
+/// Soon this will throw if any unexpected arguments are passed.
+BuilderFactory noArgs(Builder builder) => (List<String> args) {
+      if (args?.isNotEmpty ?? false) {
+        throw new StateError('No arguments should be passed to this builder.');
+      }
+      return builder;
+    };
+
+/// A step in the build, defining a [Builder], an input extension, and a
+/// list of output extensions.
+///
+/// Multiple phases run sequentially.
+class Phase {
```

Comment:
This is one of the things I would hope we would resolve in this port or right after :) +1